### PR TITLE
Document the Unification of tools/target build type configuration

### PR DESCRIPTION
**Pull request by NeroBurner**
Introduced in https://github.com/godotengine/godot/pull/66242 the `tools=yes/no` option was removed and merged into the `target` preset.

Replace the `tools=yes/no` calls in the documentation to the new template target presets `editor`, `template_release` and `template_debug`.

---

It's my first PR here, so please tell me if something is missing or done wrong. I just was confused why the `scons target=debug` builds didn't work. After finding out about the change I thought I'd take this opportunity to fix the documentation for everyone and learn quite a bit in the process

### CODE REVIEW

File: development/compiling/compiling_for_android.rst
Comment (bruvzg) [inline development/compiling/compiling_for_android.rst:97]: `android_arch` no longer exist, it's `arch` now (on all platforms).

File: development/compiling/compiling_for_linuxbsd.rst
Comment (bruvzg) [inline development/compiling/compiling_for_linuxbsd.rst:200]: `bits` no longer exist, `arch` is used instead.

File: development/compiling/compiling_for_macos.rst
Comment (bruvzg) [inline development/compiling/compiling_for_macos.rst:103]: `server` platform no longer exist, it's replaced with `--headless` command line argument of the normal executable.

File: development/compiling/compiling_with_mono.rst
Comment (bruvzg) [inline development/compiling/compiling_with_mono.rst:63]: `mono_glue` is no longer used. Build process is:
- Build executable with `module_mono_enabled=yes`
- Run ` --headless --generate-mono-glue modules/mono/glue`
- Run `./modules/mono/build_scripts/build_assemblies.py --godot-output-dir=./bin --godot-platform={PLATFORM_NAME}`

File: development/compiling/compiling_with_mono.rst
Comment (bruvzg) [inline development/compiling/compiling_with_mono.rst:80]: File names use `arch` instead `bits` as well.

File: development/compiling/optimizing_for_size.rst
Comment (bruvzg) [inline development/compiling/optimizing_for_size.rst:170]: The list of modules seems to be outdated.

File: development/compiling/compiling_for_linuxbsd.rst
Comment (gamedevsam) [inline development/compiling/compiling_for_linuxbsd.rst:152]: Shouldn't we also use `production=yes` here?

**Review by gamedevsam — CHANGES_REQUESTED** [review]
The docs for setting up Visual Studio Code are also incorrect at this point. I was unable to hit any breakpoints until I set the `"dev_build=yes"` scons option in my build task.

File: development/compiling/optimizing_for_size.rst
Comment (NeroBurner) [reply development/compiling/optimizing_for_size.rst:170]: where can I get an updated list? I don't have access to a WIndows machine to experiment

File: development/compiling/compiling_for_linuxbsd.rst
Comment (NeroBurner) [reply development/compiling/compiling_for_linuxbsd.rst:200]: on linux the bits field is replaced with `x86_64`, is that the same on other platforms (like windows)?

```suggestion
``bin/godot.<platform>.editor.<arch>.mono``, e.g. ``bin/godot.linuxbsd.editor.x86_64.mono``
```

File: development/compiling/compiling_with_mono.rst
Comment (NeroBurner) [reply development/compiling/compiling_with_mono.rst:80]: on linux the bits field is replaced with `x86_64`, is that the same on other platforms (like windows)?

```suggestion
``bin/godot.<platform>.editor.<arch>.mono``, e.g. ``bin/godot.linuxbsd.editor.x86_64.mono``
```

File: development/compiling/compiling_for_android.rst
Comment (NeroBurner) [reply development/compiling/compiling_for_android.rst]: updated, thanks

File: development/compiling/compiling_for_macos.rst
Comment (NeroBurner) [reply development/compiling/compiling_for_macos.rst]: I tried to update the section with minimal changes, please review the wording

File: development/compiling/compiling_for_linuxbsd.rst
Comment (NeroBurner) [reply development/compiling/compiling_for_linuxbsd.rst]: added it here

File: development/compiling/compiling_for_linuxbsd.rst
Comment (NeroBurner) [reply development/compiling/compiling_for_linuxbsd.rst]: i've updated every `bits=32` or `bits=64` occurrence I could find

File: development/compiling/compiling_with_mono.rst
Comment (NeroBurner) [reply development/compiling/compiling_with_mono.rst:63]: does the first binary need to be a `target=editor` build or can it be one of the `template_debug` or `template_release` templates as well?

Do we need special steps to create `target=template_release production=yes` builds?

In the current documentation examples list all the commands to create all three variants. What would be the updated equivalent?

```sh
    # Build temporary binary
    scons p=windows target=editor module_mono_enabled=yes
    # Generate glue sources
    bin\godot.windows.editor.x86_64.mono --generate-mono-glue modules/mono/glue
    # Build binaries
    ./modules/mono/build_scripts/build_assemblies.py --godot-output-dir=./bin --godot-platform={PLATFORM_NAME}

    # are the below steps needed with the new commands?
    ### Build binaries normally
    # Editor
    scons p=windows target=editor module_mono_enabled=yes
    # Export templates
    scons p=windows target=template_debug module_mono_enabled=yes
    scons p=windows target=template_release module_mono_enabled=yes
```

**Comment by NeroBurner** [comment]
I've tried to incorporate all feedback. Some things are still unclear (see above). Please re-review

File: development/compiling/optimizing_for_size.rst
Comment (NeroBurner) [reply development/compiling/optimizing_for_size.rst:170]: I've created a new list with the following command on linux. Would this be the right list?

```sh
scons --help | grep enabled | cut -d ":" -f 1 | xargs printf " %s=no"
```

```txt
modules_enabled_by_default=no module_basis_universal_enabled=no module_bmp_enabled=no module_camera_enabled=no module_csg_enabled=no module_cvtt_enabled=no module_dds_enabled=no module_denoise_enabled=no module_enet_enabled=no module_etcpak_enabled=no module_freetype_enabled=no module_gdscript_enabled=no module_glslang_enabled=no module_gltf_enabled=no module_gridmap_enabled=no module_hdr_enabled=no module_jpg_enabled=no module_jsonrpc_enabled=no module_lightmapper_rd_enabled=no module_mbedtls_enabled=no module_meshoptimizer_enabled=no module_minimp3_enabled=no module_mobile_vr_enabled=no module_mono_enabled=no module_msdfgen_enabled=no module_multiplayer_enabled=no module_navigation_enabled=no module_noise_enabled=no module_ogg_enabled=no module_openxr_enabled=no module_raycast_enabled=no module_regex_enabled=no module_squish_enabled=no module_svg_enabled=no module_text_server_adv_enabled=no module_text_server_fb_enabled=no module_tga_enabled=no module_theora_enabled=no module_tinyexr_enabled=no module_upnp_enabled=no module_vhacd_enabled=no module_vorbis_enabled=no module_webp_enabled=no module_webrtc_enabled=no module_websocket_enabled=no module_webxr_enabled=no module_xatlas_unwrap_enabled=no
```

File: development/cpp/configuring_an_ide/visual_studio_code.rst
Comment (Maran23) [inline development/cpp/configuring_an_ide/visual_studio_code.rst:67]: the image below also needs to be adjusted. (https://docs.godotengine.org/en/latest/_images/vscode_3_tasks.json.png)
If you can't, I can do that in a follow up since I'm using VSCode anyway. :)

Comment (NeroBurner) [reply development/cpp/configuring_an_ide/visual_studio_code.rst:67]: Updated, just needed to figure out how to toggle dark-mode :grin:

**Comment by NeroBurner** [comment]
rebased to solve conflicts

File: development/cpp/configuring_an_ide/visual_studio_code.rst
Comment (Maran23) [inline development/cpp/configuring_an_ide/visual_studio_code.rst:149]: There is one more screenshot below which needs to be changed :):
https://docs.godotengine.org/en/latest/_images/vscode_2_launch.json.png

File: development/compiling/compiling_with_mono.rst
Comment (Maran23) [reply development/compiling/compiling_with_mono.rst]: Yes.

File: development/compiling/compiling_with_mono.rst
Comment (raulsntos) [inline development/compiling/compiling_with_mono.rst:377]: `mono_prefix` is now an argument of the `build_assemblies.py` script, although it probably should be removed as we no longer use Mono and I feel like using Mono's MSBuild may not work.
`mono_static` and `copy_mono_root` have also been removed in 4.0.

**Review by raulsntos — COMMENTED** [review]
I looked at the instructions for compiling with "Mono" and left a few suggestions but there are many other issues unrelated to this PR with this documentation since building with the .NET module is very different from what it was in 3.x.

Mono is no longer required as it's not used anymore, we also no longer build the Mono runtime and mobile is not currently supported (likely won't be until 4.1+). Since this grows the scope of the PR I think it's fine to ignore and open a separate PR for the .NET documentation.

Keep in mind that the process will likely change again with the editor unification so it may not be worth updating the .NET documentation for now.

File: development/cpp/configuring_an_ide/visual_studio_code.rst
Comment (NeroBurner) [reply development/cpp/configuring_an_ide/visual_studio_code.rst:149]: updated, thanks for the hint

**Comment by Maran23** [comment]
> I looked at the instructions for compiling with "Mono" and left a few suggestions but there are many other issues unrelated to this PR with this documentation since building with the .NET module is very different from what it was in 3.x.
> 
> Mono is no longer required as it's not used anymore, we also no longer build the Mono runtime and mobile is not currently supported (likely won't be until 4.1+). Since this grows the scope of the PR I think it's fine to ignore and open a separate PR for the .NET documentation.
> 
> Keep in mind that the process will likely change again with the editor unification so it may not be worth updating the .NET documentation for now.

I'm in favor of merging this PR first and doing that work in a follow-up PR. 
I wasted like an hour building the Godot 4.0 project because the documentation does not reflect the changes from the https://github.com/godotengine/godot/pull/66242 PR. 
We should avoid that, I almost gave up because I didn't know what to do (it worked for me before obviously and I thought the whole time it's an issue with my VSCode).

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (Maran23) [inline development/compiling/introduction_to_the_buildsystem.rst:150]:
```suggestion
When doing engine development the ``dev_build`` option can be used together
```

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (Maran23) [inline development/compiling/introduction_to_the_buildsystem.rst:148]: Shouldn't we still explain the different targets?
- editor
- template_debug
- template_release

File: development/cpp/configuring_an_ide/visual_studio_code.rst
Comment (Maran23) [inline development/cpp/configuring_an_ide/visual_studio_code.rst:176]:
```suggestion
e.g. ``godot.linuxbsd.editor.dev.x86_64`` for 64-bit LinuxBSD platform with
```

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (Maran23) [inline development/compiling/introduction_to_the_buildsystem.rst:175]: The default value ``arch=auto`` detects the architecture that matches the host platform. 
Sounds a bit better here.

File: development/compiling/compiling_for_macos.rst
Comment (Maran23) [inline development/compiling/compiling_for_macos.rst:122]: ... to compile using the targets without the
editor: ``target=template_release`` (release template) and
``target=template_debug

**Comment by Maran23** [comment]
I left some comments regarding spelling and typos. Otherwise it looks fine and I want to get this merged!:) 
Also it is mandatory to squash the commits together as also stated here: https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#modifying-a-pull-request

**Comment by Calinou** [comment]
> Also it is mandatory to squash the commits together as also stated here: [docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#modifying-a-pull-request](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#modifying-a-pull-request)

We can use *Squash on merge* on the godot-docs repository, so it's not necessary here.

File: development/compiling/compiling_with_mono.rst
Comment (Calinou) [reply development/compiling/compiling_with_mono.rst:138]: Forward slash directory separators should be converted automatically by Windows, but the leading `./` should be removed (it can also be removed on macOS/Linux).

**Review by Maran23 — APPROVED** [review]
Looks good to me.
The Mono/.NET documentation should be updated in a follow up and was partly outdated before already.

**Review by Maran23 — APPROVED** [review]
LGTM

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (clayjohn) [inline development/compiling/introduction_to_the_buildsystem.rst:101]:
```suggestion
    godot.<platform>.<target>[.dev][.double].<arch>[.<extra_suffix>][.<ext>]
```

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (clayjohn) [inline development/compiling/introduction_to_the_buildsystem.rst:132]:
```suggestion
Target controls if the editor is contained and debug flags are used.
```

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (clayjohn) [inline development/compiling/introduction_to_the_buildsystem.rst:135]:
```suggestion
-  **editor**: Build with editor, optimized, with debugging code (defines: `TOOLS_ENABLED`, `DEBUG_ENABLED`, `-O2`/`/O2`)
```

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (clayjohn) [inline development/compiling/introduction_to_the_buildsystem.rst:137]:
```suggestion
-  **template_debug**: Build with C++ debugging symbols (defines: `DEBUG_ENABLED`, `-O2`/`/O2`)
-  **template_release**: Build without symbols (defines: `-O3`/`/O2`)
```

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (clayjohn) [inline development/compiling/introduction_to_the_buildsystem.rst:151]:
```suggestion
When doing engine development the ``dev_build`` option can be used together
with ``target`` to enable dev-specific code. ``dev_build`` defines ``DEV_ENABLED``,
disables optimization (``-O0``/``/0d``), enables generating debug symbols, and
does not define ``NDEBUG`` (so `assert()` works in thirdparty libraries).
```

**Review by clayjohn — COMMENTED** [review]
Mostly looks good to me. I reviewed the introduction to buildsystem page quite closely as well as the platforms I am familiar with: Android, Linuxbsd, macOS, web, and windows. 

I have a left a few suggestions for improvements. I can also see that there are a few outstanding comments for building with mono. The building with mono page seems quite complex and it may be best to have one of the mono maintainers do the update rather than including it here. I want to avoid blocking this PR any longer as these pages desperately need the update.

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (Calinou) [reply development/compiling/introduction_to_the_buildsystem.rst:137]: @clayjohn Backticks should be doubled in rSt syntax: \`\`

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (Calinou) [reply development/compiling/introduction_to_the_buildsystem.rst:135]: @clayjohn Each backtick should be doubled in reStructuredText syntax, so \`-O2\` should be \`\`-O2\`\`.

Comment (NeroBurner) [reply development/compiling/introduction_to_the_buildsystem.rst:135]:
```suggestion
-  **editor**: Build with editor, optimized, with debugging code (defines: ``TOOLS_ENABLED``, ``DEBUG_ENABLED``, ``-O2``/``/O2``)
```

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (NeroBurner) [reply development/compiling/introduction_to_the_buildsystem.rst:137]:
```suggestion
-  **template_debug**: Build with C++ debugging symbols (defines: ``DEBUG_ENABLED``, ``-O2``/``/O2``)
-  **template_release**: Build without symbols (defines: ``-O3``/``/O2``)
```

File: development/compiling/introduction_to_the_buildsystem.rst
Comment (NeroBurner) [reply development/compiling/introduction_to_the_buildsystem.rst:151]:
```suggestion
When doing engine development the ``dev_build`` option can be used together
with ``target`` to enable dev-specific code. ``dev_build`` defines ``DEV_ENABLED``,
disables optimization (``-O0``/``/0d``), enables generating debug symbols, and
does not define ``NDEBUG`` (so ``assert()`` works in thirdparty libraries).
```

**Review by clayjohn — APPROVED** [review]
Everything except the mono page seem fine to me. 

We discussed this a bit on the rocketchat and agreed to move forward with this PR as-is. The Mono changes are not from https://github.com/godotengine/godot/pull/66242, but from all of the other changes that have happened recently so it is fine for them to be made in a separate PR from someone who is more knowledgeable about the c sharp specific changes to the buildsystem

**Review by Maran23 — APPROVED** [review]
Thanks!

**Comment by clayjohn** [comment]
Thank you very much @NeroBurner for your hard work and sticking with this PR. And congratulations on your first merged PR in this repo!

Thanks to you as well @Maran23 for your multiple diligent reviews.

**Comment by Anutrix** [comment]
Was `dev_mode` intentionally left out?

**Pull request merged**
